x/wasm/types/authz_test.go

```diff
@@ -744,3 +745,184 @@ func mustGrant(contract sdk.AccAddress, limit ContractAuthzLimitX, filter Contra
 	}
 	return *g
 }
+
+func TestValidateCodeGrant(t *testing.T) {
+	specs := map[string]struct {
+		codeHash              []byte
+		instantiatePermission AccessConfig
+		expErr                bool
+	}{
+		"all good": {
+			codeHash:              []byte("ABC"),
+			instantiatePermission: AllowEverybody,
+		},
+		"empty code hash": {
+			codeHash:              []byte{},
+			instantiatePermission: AllowEverybody,
+			expErr:                true,
+		},
+		"nil code hash": {
+			codeHash:              nil,
+			instantiatePermission: AllowEverybody,
+			expErr:                true,
+		},
+		"invalid permission": {
+			codeHash:              []byte("ABC"),
+			instantiatePermission: AccessConfig{Permission: AccessTypeUnspecified},
+			expErr:                true,
+		},
+	}
+	for name, spec := range specs {
+		t.Run(name, func(t *testing.T) {
+			grant, err := NewCodeGrant(spec.codeHash, spec.instantiatePermission)
+			require.NoError(t, err)
+
+			gotErr := grant.ValidateBasic()
+			if spec.expErr {
+				require.Error(t, gotErr)
+				return
+			}
+			require.NoError(t, gotErr)
+		})
+	}
+}
+
+func TestValidateStoreCodeAuthorization(t *testing.T) {
+	validGrant, err := NewCodeGrant([]byte("ABC"), AllowEverybody)
+	require.NoError(t, err)
+	invalidGrant, err := NewCodeGrant(nil, AllowEverybody)
+	require.NoError(t, err)
+
+	specs := map[string]struct {
+		setup  func(t *testing.T) validatable
+		expErr bool
+	}{
+		"all good": {
+			setup: func(t *testing.T) validatable {
+				t.Helper()
+				return NewStoreCodeAuthorization(*validGrant)
+			},
+		},
+		"duplicate grants": {
+			setup: func(t *testing.T) validatable {
+				t.Helper()
+				return NewStoreCodeAuthorization(*validGrant, *validGrant)
+			},
+		},
+		"invalid grant": {
+			setup: func(t *testing.T) validatable {
+				t.Helper()
+				return NewStoreCodeAuthorization(*validGrant, *invalidGrant)
+			},
+			expErr: true,
+		},
+		"empty grants": {
+			setup: func(t *testing.T) validatable {
+				t.Helper()
+				return NewStoreCodeAuthorization()
+			},
+			expErr: true,
+		},
```

Comment:
Duplicate grants can be more complex. Please add more cases (and doc) to ensure that this is well defined:
` *, AllowEverybody` and `ABC, AllowEverybody` must not be allowed for example
Not sure about `ABC, AllowEverybody` and `ABC, unset` or other cases where only the accessConfig differs.
There is also the double wildcard with `*,unset` to consider